File: xxcxy/RemoveLinkedListElements.java
/**
 * Definition for singly-linked list.
*/
 class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}
 
public class RemoveLinkedListElements {
    /**
     * @param head a ListNode
     * @param val an integer
     * @return a ListNode
     */
    public ListNode removeElements(ListNode head, int val) {
        head = toConformLocation(head, val);
        if(head == null) {
            return null;
        }
        ListNode outcome = new ListNode(head.val);
        ListNode next = outcome;
        while(head.next != null) {
            if(head.next.val != val){
                next = appendNode(next, head.next.val);
            }
            head = head.next;
        }
        return outcome;
    }
    
    private ListNode toConformLocation(ListNode head, int val) {
        while(head !=null && head.val == val) {
            head = head.next;
        }
        return head;
    }
    
    private ListNode appendNode(ListNode last, int val) {
        ListNode current = new ListNode(val);
        if(last != null) {
            last.next = current;
        }
        return current;
    }
}

File: lulu/remove_linked_list_elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    # @param head, a ListNode
    # @param val, an integer
    # @return a ListNode

    # Non-recursive way
    def removeElements(self, head, val):
        # Write your code here
        i = head
        while i is not None:
            if i.val != val:
                head = i
                break
            i = i.next
        if i is None:
            return None
        former = head
        i = head.next
        while i is not None:
            if i.val == val:
                former.next = i.next
            else:
                former = i
            i = i.next
        return head

    # Recursive way
    def removeElements(self, head, val):
        if head is None:
            return None
        if head.val == val:
            return self.removeElements(head.next, val)
        head.next = self.removeElements(head.next, val)
        return head
